### PR TITLE
Allow throwing exceptions for HTTP error codes to better check query param input.

### LINKED ISSUE #13752: fatal error triggered under some circumstances
The fatal error: `Neither the property "getRows" nor one of the methods "getRows()", "getgetRows()"/"isgetRows()" or "__call()" exist and have public access in class "Piwik\DataTable\Map".
in plugins/CoreVisualizations/templates/_dataTableViz_htmlTable.twig line 21`

Can be reproduced at https://demo.matomo.org/index.php?forceView=1&viewDataTable=..%2fWEB-INF%2fweb.xml%3bx%3d&module=MediaAnalytics&action=get&idSite=3&period=day&date=2018-10-16,2018-11-14&showtitle=1&random=84&columns=sum_total_time_watched&colors={%22backgroundColor%22%3A%22%23ffffff%22%2C%22lineColor%22%3A%22%23162c4a%22%2C%22minPointColor%22%3A%22%23ff7f7f%22%2C%22maxPointColor%22%3A%22%2375bf7c%22%2C%22lastPointColor%22%3A%22%2355aaff%22%2C%22fillColor%22%3A%22%23ffffff%22}

It is only triggered under edge case (ie. security scanner), but will be good to fix since we get this emailed to us quite a bit.

### PR DESCRIPTION
**Pull request by diosmosis**
Added two new exceptions `HttpCodeException` and `BadRequestException` that allows changing the HTTP code returned for an exception (in both controller & API).

This allows us to avoid certain failures caused by invalid input and correctly report the error, w/o triggering a 500 HTTP code (that could potentially also create an email alert?).

Used in ScheduledReports API to check for multiple periods.

Fixes #13752

### CODE REVIEW

File: plugins/CoreVisualizations/Visualizations/HtmlTable.php [hunk 1]
```diff
@@ -10,6 +10,7 @@
 
 use Piwik\API\Request as ApiRequest;
 use Piwik\Common;
+use Piwik\Http\BadRequestException;
```
Comment (sgiehl) [inline plugins/CoreVisualizations/Visualizations/HtmlTable.php]: guess that could be removed again

**Review by sgiehl — CHANGES_REQUESTED** [review]

**Comment by sgiehl** [comment]
@diosmosis  this one needs a rebase / merge to solve the conflicts

File: core/API/ResponseBuilder.php [hunk 2]
```diff
@@ -135,6 +136,10 @@ public function getResponseException($e)
         $e       = $this->decorateExceptionWithDebugTrace($e);
         $message = $this->formatExceptionMessage($e);
 
+        if ($this->sendHeader && $e instanceof HttpCodeException) {
```
Comment (tsteur) [inline core/API/ResponseBuilder.php]: would you need to check if a code is set?

File: core/ExceptionHandler.php [hunk 3]
```diff
@@ -70,6 +71,10 @@ public static function dieWithCliError($exception)
      */
     public static function dieWithHtmlErrorPage($exception)
     {
+        if ($exception instanceof HttpCodeException) {
```
Comment (tsteur) [inline core/ExceptionHandler.php]: same here maybe we better check if a code is set?

**Comment by tsteur** [comment]
Left 2 comments, otherwise good to merge when tests pass 👍

**Pull request merged**